File: idea-plugin/src/main/java/com/lingce/cleaner/AiCleanerConfigurable.java
package com.lingce.cleaner;

import com.intellij.openapi.options.Configurable;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.Nullable;

import javax.swing.JCheckBox;
import javax.swing.JComponent;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JPasswordField;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import java.awt.BorderLayout;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.util.Objects;

public final class AiCleanerConfigurable implements Configurable {
    private JPanel panel;
    private JCheckBox enabled;
    private JCheckBox scanOnProjectOpen;
    private JCheckBox watchNewFiles;
    private JCheckBox remoteDetectionEnabled;
    private JCheckBox autoDeleteTempFiles;
    private JCheckBox autoDeleteDisposableAiFiles;
    private JCheckBox autoIgnoreConfigFiles;
    private JTextField baseUrl;
    private JPasswordField apiKey;
    private JTextField model;
    private JTextField quarantineDirectory;
    private JTextField ignoreFileName;
    private JTextField maxScanFileSizeKb;
    private JTextArea customTempPatterns;

    @Override
    public @Nls(capitalization = Nls.Capitalization.Title) String getDisplayName() {
        return "LingCe AI File Cleaner";
    }

    @Override
    public @Nullable JComponent createComponent() {
        panel = new JPanel(new GridBagLayout());

        enabled = new JCheckBox("Enable LingCe AI File Cleaner");
        scanOnProjectOpen = new JCheckBox("Scan project when it opens");
        watchNewFiles = new JCheckBox("Watch and process newly created files");
        remoteDetectionEnabled = new JCheckBox("Use OpenAI-compatible API for uncertain files");
        autoDeleteTempFiles = new JCheckBox("Automatically delete temporary files");
        autoDeleteDisposableAiFiles = new JCheckBox("Automatically delete disposable AI-generated files");
        autoIgnoreConfigFiles = new JCheckBox("Add project/AI configuration files to ignore rules");
        baseUrl = new JTextField();
        apiKey = new JPasswordField();
        model = new JTextField();
        quarantineDirectory = new JTextField();
        ignoreFileName = new JTextField();
        maxScanFileSizeKb = new JTextField();
        customTempPatterns = new JTextArea(5, 48);

        int row = 0;
        addFullWidth(row++, enabled);
        addFullWidth(row++, scanOnProjectOpen);
        addFullWidth(row++, watchNewFiles);
        addFullWidth(row++, remoteDetectionEnabled);
        addLabeled(row++, "Base URL", baseUrl);
        addLabeled(row++, "API Key", apiKey);
        addLabeled(row++, "Model", model);
        addFullWidth(row++, autoDeleteTempFiles);
        addFullWidth(row++, autoDeleteDisposableAiFiles);
        addFullWidth(row++, autoIgnoreConfigFiles);
        addLabeled(row++, "Quarantine directory", quarantineDirectory);
        addLabeled(row++, "Ignore file name", ignoreFileName);
        addLabeled(row++, "Max scanned file size (KB)", maxScanFileSizeKb);
        addFullWidth(row++, new JLabel("Custom temporary file regex patterns, one per line:"));
        addFullWidth(row, new JScrollPane(customTempPatterns));

        reset();
        return panel;
    }

    @Override
    public boolean isModified() {
        AiCleanerSettings.StateData state = AiCleanerSettings.getInstance().getState();
        return enabled.isSelected() != state.enabled
            || scanOnProjectOpen.isSelected() != state.scanOnProjectOpen
            || watchNewFiles.isSelected() != state.watchNewFiles
            || remoteDetectionEnabled.isSelected() != state.remoteDetectionEnabled
            || autoDeleteTempFiles.isSelected() != state.autoDeleteTempFiles
            || autoDeleteDisposableAiFiles.isSelected() != state.autoDeleteDisposableAiFiles
            || autoIgnoreConfigFiles.isSelected() != state.autoIgnoreConfigFiles
            || !Objects.equals(baseUrl.getText(), state.baseUrl)
            || !Objects.equals(new String(apiKey.getPassword()), state.apiKey)
            || !Objects.equals(model.getText(), state.model)
            || !Objects.equals(quarantineDirectory.getText(), state.quarantineDirectory)
            || !Objects.equals(ignoreFileName.getText(), state.ignoreFileName)
            || !Objects.equals(maxScanFileSizeKb.getText(), Integer.toString(state.maxScanFileSizeKb))
            || !Objects.equals(customTempPatterns.getText(), state.customTempPatterns);
    }

    @Override
    public void apply() {
        AiCleanerSettings.StateData state = AiCleanerSettings.getInstance().getState();
        state.enabled = enabled.isSelected();
        state.scanOnProjectOpen = scanOnProjectOpen.isSelected();
        state.watchNewFiles = watchNewFiles.isSelected();
        state.remoteDetectionEnabled = remoteDetectionEnabled.isSelected();
        state.autoDeleteTempFiles = autoDeleteTempFiles.isSelected();
        state.autoDeleteDisposableAiFiles = autoDeleteDisposableAiFiles.isSelected();
        state.autoIgnoreConfigFiles = autoIgnoreConfigFiles.isSelected();
        state.baseUrl = baseUrl.getText().trim();
        state.apiKey = new String(apiKey.getPassword()).trim();
        state.model = model.getText().trim();
        state.quarantineDirectory = quarantineDirectory.getText().trim();
        state.ignoreFileName = ignoreFileName.getText().trim();
        state.maxScanFileSizeKb = parsePositiveInt(maxScanFileSizeKb.getText(), state.maxScanFileSizeKb);
        state.customTempPatterns = customTempPatterns.getText();
    }

    @Override
    public void reset() {
        AiCleanerSettings.StateData state = AiCleanerSettings.getInstance().getState();
        enabled.setSelected(state.enabled);
        scanOnProjectOpen.setSelected(state.scanOnProjectOpen);
        watchNewFiles.setSelected(state.watchNewFiles);
        remoteDetectionEnabled.setSelected(state.remoteDetectionEnabled);
        autoDeleteTempFiles.setSelected(state.autoDeleteTempFiles);
        autoDeleteDisposableAiFiles.setSelected(state.autoDeleteDisposableAiFiles);
        autoIgnoreConfigFiles.setSelected(state.autoIgnoreConfigFiles);
        baseUrl.setText(state.baseUrl);
        apiKey.setText(state.apiKey);
        model.setText(state.model);
        quarantineDirectory.setText(state.quarantineDirectory);
        ignoreFileName.setText(state.ignoreFileName);
        maxScanFileSizeKb.setText(Integer.toString(state.maxScanFileSizeKb));
        customTempPatterns.setText(state.customTempPatterns);
    }

    @Override
    public void disposeUIResources() {
        panel = null;
    }

    private void addLabeled(int row, String label, JComponent component) {
        GridBagConstraints labelConstraints = new GridBagConstraints();
        labelConstraints.gridx = 0;
        labelConstraints.gridy = row;
        labelConstraints.anchor = GridBagConstraints.WEST;
        labelConstraints.insets = new Insets(4, 0, 4, 8);
        panel.add(new JLabel(label + ":"), labelConstraints);

        GridBagConstraints fieldConstraints = new GridBagConstraints();
        fieldConstraints.gridx = 1;
        fieldConstraints.gridy = row;
        fieldConstraints.weightx = 1;
        fieldConstraints.fill = GridBagConstraints.HORIZONTAL;
        fieldConstraints.insets = new Insets(4, 0, 4, 0);
        panel.add(component, fieldConstraints);
    }

    private void addFullWidth(int row, JComponent component) {
        JPanel wrapper = new JPanel(new BorderLayout());
        wrapper.add(component, BorderLayout.CENTER);

        GridBagConstraints constraints = new GridBagConstraints();
        constraints.gridx = 0;
        constraints.gridy = row;
        constraints.gridwidth = 2;
        constraints.weightx = 1;
        constraints.fill = GridBagConstraints.HORIZONTAL;
        constraints.anchor = GridBagConstraints.WEST;
        constraints.insets = new Insets(4, 0, 4, 0);
        panel.add(wrapper, constraints);
    }

    private int parsePositiveInt(String value, int fallback) {
        try {
            int parsed = Integer.parseInt(value.trim());
            return Math.max(1, parsed);
        } catch (NumberFormatException ignored) {
            return fallback;
        }
    }
}

File: idea-plugin/src/main/kotlin/com/lingce/aicleaner/settings/AiCleanerConfigurable.kt
package com.lingce.aicleaner.settings

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.options.Configurable
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.ui.Messages
import com.intellij.ui.components.JBPasswordField
import com.intellij.ui.components.JBTextArea
import com.intellij.ui.components.JBTextField
import com.intellij.util.ui.JBUI
import com.lingce.aicleaner.core.AiClient
import javax.swing.JCheckBox
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JPanel
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.Insets

/**
 * 设置页：配置 AI 接口（baseUrl / apiKey / model）、行为开关与各类清理规则。
 */
class AiCleanerConfigurable : Configurable {

    private val settings = AiCleanerSettings.getInstance()

    private val baseUrlField = JBTextField()
    private val apiKeyField = JBPasswordField()
    private val modelField = JBTextField()
    private val quarantineField = JBTextField()

    private val enableAiBox = JCheckBox("启用 AI 辅助判定（本地规则不确定时调用）")
    private val watchBox = JCheckBox("监听导入/新建文件并自动分类提示")
    private val autoCleanBox = JCheckBox("自动执行清理（可疑文件永远不会被自动删除）")
    private val deleteToQuarantineBox = JCheckBox("删除前先转存到隔离目录（更安全）")

    private val tmpArea = JBTextArea(3, 60)
    private val aiConfigArea = JBTextArea(3, 60)
    private val projectConfigArea = JBTextArea(3, 60)
    private val aiGenNameArea = JBTextArea(3, 60)
    private val markersArea = JBTextArea(4, 60)

    override fun getDisplayName(): String = "AI 文件清理 (AI File Cleaner)"

    override fun createComponent(): JComponent {
        val panel = JPanel(GridBagLayout())
        val c = GridBagConstraints().apply {
            gridx = 0; gridy = 0
            anchor = GridBagConstraints.WEST
            fill = GridBagConstraints.HORIZONTAL
            weightx = 1.0
            insets = Insets(4, 8, 4, 8)
            gridwidth = 2
        }

        fun add(comp: JComponent) {
            panel.add(comp, c); c.gridy++
        }

        fun section(title: String) {
            val l = JLabel("<html><b>$title</b></html>")
            l.border = JBUI.Borders.emptyTop(10)
            add(l)
        }

        section("AI 接口配置")
        add(labeled("Base URL：", baseUrlField))
        add(labeled("API Key：", apiKeyField))
        add(labeled("Model：", modelField))
        val testBtn = javax.swing.JButton("测试连接")
        testBtn.addActionListener { testConnection() }
        add(testBtn)

        section("行为")
        add(enableAiBox)
        add(watchBox)
        add(autoCleanBox)
        add(deleteToQuarantineBox)
        add(labeled("隔离目录(留空=项目根/.ai-cleaner-quarantine)：", quarantineField))

        section("规则：临时文件（删除）")
        add(scroll(tmpArea))
        section("规则：AI 配置文件（加入忽略/排除）")
        add(scroll(aiConfigArea))
        section("规则：项目配置文件（加入忽略/排除）")
        add(scroll(projectConfigArea))
        section("规则：AI 生成无用文件 - 文件名（删除）")
        add(scroll(aiGenNameArea))
        section("规则：AI 生成标记 - 内容关键字，每行一个（删除）")
        add(scroll(markersArea))

        // 撑满底部
        c.weighty = 1.0; c.fill = GridBagConstraints.BOTH
        panel.add(JPanel(), c)

        reset()
        return panel
    }

    private fun labeled(label: String, field: JComponent): JPanel {
        val p = JPanel(GridBagLayout())
        val lc = GridBagConstraints().apply {
            gridx = 0; gridy = 0; anchor = GridBagConstraints.WEST
            insets = Insets(0, 0, 0, 8)
        }
        val l = JLabel(label)
        l.preferredSize = java.awt.Dimension(260, l.preferredSize.height)
        p.add(l, lc)
        lc.gridx = 1; lc.weightx = 1.0; lc.fill = GridBagConstraints.HORIZONTAL
        p.add(field, lc)
        return p
    }

    private fun scroll(area: JBTextArea): JComponent {
        area.lineWrap = true
        area.wrapStyleWord = true
        return com.intellij.ui.components.JBScrollPane(area)
    }

    private fun testConnection() {
        // 先把当前界面输入写入设置，便于测试
        apply()
        ProgressManager.getInstance().runProcessWithProgressSynchronously(
            {
                val msg = AiClient(settings).testConnection()
                ApplicationManager.getApplication().invokeLater {
                    Messages.showInfoMessage(msg, "AI 连接测试")
                }
            },
            "测试 AI 连接…", true, null,
        )
    }

    override fun isModified(): Boolean {
        val s = settings.state
        return baseUrlField.text != s.baseUrl ||
            String(apiKeyField.password) != settings.apiKey ||
            modelField.text != s.model ||
            quarantineField.text != s.quarantineDir ||
            enableAiBox.isSelected != s.enableAi ||
            watchBox.isSelected != s.watchImportedFiles ||
            autoCleanBox.isSelected != s.autoClean ||
            deleteToQuarantineBox.isSelected != s.deleteToQuarantine ||
            tmpArea.text != s.tmpPatterns ||
            aiConfigArea.text != s.aiConfigPatterns ||
            projectConfigArea.text != s.projectConfigPatterns ||
            aiGenNameArea.text != s.aiGeneratedNamePatterns ||
            markersArea.text != s.aiContentMarkers
    }

    override fun apply() {
        val s = settings.state
        s.baseUrl = baseUrlField.text.trim()
        settings.apiKey = String(apiKeyField.password)
        s.model = modelField.text.trim()
        s.quarantineDir = quarantineField.text.trim()
        s.enableAi = enableAiBox.isSelected
        s.watchImportedFiles = watchBox.isSelected
        s.autoClean = autoCleanBox.isSelected
        s.deleteToQuarantine = deleteToQuarantineBox.isSelected
        s.tmpPatterns = tmpArea.text
        s.aiConfigPatterns = aiConfigArea.text
        s.projectConfigPatterns = projectConfigArea.text
        s.aiGeneratedNamePatterns = aiGenNameArea.text
        s.aiContentMarkers = markersArea.text
    }

    override fun reset() {
        val s = settings.state
        baseUrlField.text = s.baseUrl
        apiKeyField.text = settings.apiKey
        modelField.text = s.model
        quarantineField.text = s.quarantineDir
        enableAiBox.isSelected = s.enableAi
        watchBox.isSelected = s.watchImportedFiles
        autoCleanBox.isSelected = s.autoClean
        deleteToQuarantineBox.isSelected = s.deleteToQuarantine
        tmpArea.text = s.tmpPatterns
        aiConfigArea.text = s.aiConfigPatterns
        projectConfigArea.text = s.projectConfigPatterns
        aiGenNameArea.text = s.aiGeneratedNamePatterns
        markersArea.text = s.aiContentMarkers
    }
}
